File: WWF/Tutorials/Tutorial1/Exercise2/Task1/VB/SimpleExpenseReport/Program.vb
'---------------------------------------------------------------------
'  This file is part of the Windows Workflow Foundation SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'---------------------------------------------------------------------
Imports Microsoft.VisualBasic
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms
Imports System.Collections.Generic
Imports System.Workflow.Runtime
Imports System.Workflow.Runtime.Hosting
Imports System.Workflow.Activities
Imports System.Threading

Namespace Microsoft.Samples.Workflow.Tutorials.SequentialWorkflow
    Public Class MainForm : Inherits Form
        Private label1 As System.Windows.Forms.Label
        Private result As System.Windows.Forms.TextBox
        Private label2 As System.Windows.Forms.Label
        Private WithEvents submitButton As System.Windows.Forms.Button
        Private approvalState As System.Windows.Forms.Label
        Private WithEvents approveButton As System.Windows.Forms.Button
        Private WithEvents rejectButton As System.Windows.Forms.Button
        Private WithEvents amount As System.Windows.Forms.TextBox
        Private panel1 As System.Windows.Forms.Panel

        Private components As System.ComponentModel.IContainer = Nothing

        Private workflowRuntime As WorkflowRuntime = Nothing
        Private workflowInstance As WorkflowInstance = Nothing

        Public Sub New()
            InitializeComponent()

            ' Collapse approve/reject panel
            Me.Height = Me.MinimumSize.Height

            Me.workflowRuntime = New WorkflowRuntime()
            workflowRuntime.StartRuntime()

            AddHandler workflowRuntime.WorkflowCompleted, _
                AddressOf Me.workflowRuntime_WorkflowCompleted
        End Sub

        Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
            If disposing AndAlso (Not components Is Nothing) Then
                components.Dispose()
            End If
            MyBase.Dispose(disposing)
        End Sub

        Private Sub InitializeComponent()
            Me.label1 = New System.Windows.Forms.Label()
            Me.result = New System.Windows.Forms.TextBox()
            Me.label2 = New System.Windows.Forms.Label()
            Me.submitButton = New System.Windows.Forms.Button()
            Me.approvalState = New System.Windows.Forms.Label()
            Me.approveButton = New System.Windows.Forms.Button()
            Me.rejectButton = New System.Windows.Forms.Button()
            Me.amount = New System.Windows.Forms.TextBox()
            Me.panel1 = New System.Windows.Forms.Panel()
            Me.panel1.SuspendLayout()
            Me.SuspendLayout()
            ' 
            ' label1
            ' 
            Me.label1.AutoSize = True
            Me.label1.Location = New System.Drawing.Point(10, 13)
            Me.label1.Name = "label1"
            Me.label1.Size = New System.Drawing.Size(43, 13)
            Me.label1.TabIndex = 1
            Me.label1.Text = "Amount"
            ' 
            ' result
            ' 
            Me.result.Location = New System.Drawing.Point(13, 70)
            Me.result.Name = "result"
            Me.result.ReadOnly = True
            Me.result.Size = New System.Drawing.Size(162, 20)
            Me.result.TabIndex = 1
            Me.result.TabStop = False
            ' 
            ' label2
            ' 
            Me.label2.AutoSize = True
            Me.label2.Location = New System.Drawing.Point(10, 54)
            Me.label2.Name = "label2"
            Me.label2.Size = New System.Drawing.Size(37, 13)
            Me.label2.TabIndex = 3
            Me.label2.Text = "Result"
            ' 
            ' submitButton
            ' 
            Me.submitButton.Enabled = False
            Me.submitButton.Location = New System.Drawing.Point(56, 95)
            Me.submitButton.Name = "submitButton"
            Me.submitButton.Size = New System.Drawing.Size(75, 23)
            Me.submitButton.TabIndex = 2
            Me.submitButton.Text = "Submit"
            ' 
            ' approvalState
            ' 
            Me.approvalState.AutoSize = True
            Me.approvalState.Location = New System.Drawing.Point(10, 9)
            Me.approvalState.Name = "approvalState"
            Me.approvalState.Size = New System.Drawing.Size(49, 13)
            Me.approvalState.TabIndex = 4
            Me.approvalState.Text = "Approval"
            ' 
            ' approveButton
            ' 
            Me.approveButton.Enabled = False
            Me.approveButton.Location = New System.Drawing.Point(11, 25)
            Me.approveButton.Name = "approveButton"
            Me.approveButton.Size = New System.Drawing.Size(75, 23)
            Me.approveButton.TabIndex = 0
            Me.approveButton.Text = "Approve"
            ' 
            ' rejectButton
            ' 
            Me.rejectButton.Enabled = False
            Me.rejectButton.Location = New System.Drawing.Point(86, 25)
            Me.rejectButton.Name = "rejectButton"
            Me.rejectButton.Size = New System.Drawing.Size(75, 23)
            Me.rejectButton.TabIndex = 1
            Me.rejectButton.Text = "Reject"
            ' 
            ' amount
            ' 
            Me.amount.Location = New System.Drawing.Point(13, 29)
            Me.amount.MaxLength = 9
            Me.amount.Name = "amount"
            Me.amount.Size = New System.Drawing.Size(162, 20)
            Me.amount.TabIndex = 0
            ' 
            ' panel1
            ' 
            Me.panel1.Controls.Add(Me.approvalState)
            Me.panel1.Controls.Add(Me.approveButton)
            Me.panel1.Controls.Add(Me.rejectButton)
            Me.panel1.Location = New System.Drawing.Point(7, 124)
            Me.panel1.Name = "panel1"
            Me.panel1.Size = New System.Drawing.Size(172, 66)
            Me.panel1.TabIndex = 8
            ' 
            ' MainForm
            ' 
            Me.AcceptButton = Me.submitButton
            Me.AutoScaleDimensions = New System.Drawing.SizeF(6.0F, 13.0F)
            Me.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Font
            Me.ClientSize = New System.Drawing.Size(187, 201)
            Me.MinimumSize = New System.Drawing.Size(197,155)
            Me.Controls.Add(Me.result)
            Me.Controls.Add(Me.label2)
            Me.Controls.Add(Me.panel1)
            Me.Controls.Add(Me.amount)
            Me.Controls.Add(Me.submitButton)
            Me.Controls.Add(Me.label1)
            Me.FormBorderStyle = System.Windows.Forms.FormBorderStyle.Fixed3D
            Me.MaximizeBox = False
            Me.MinimizeBox = False
            Me.Name = "MainForm"
            Me.Text = "Simple Expense Report"
            Me.panel1.ResumeLayout(False)
            Me.panel1.PerformLayout()
            Me.ResumeLayout(False)
            Me.PerformLayout()
        End Sub

        Private Sub submitButton_Click(ByVal sender As Object, _
                  ByVal e As EventArgs) Handles submitButton.Click
            Dim type As Type = GetType(ExpenseReportWorkflow)

            ' Construct workflow parameters
            Dim properties As Dictionary(Of String, Object) = _
                New Dictionary(Of String, Object)()
            properties.Add("Amount", Int32.Parse(Me.amount.Text))

            ' Start the workflow
            workflowInstance = workflowRuntime.CreateWorkflow(type, properties)
            workflowInstance.Start()
        End Sub

        Private Sub workflowRuntime_WorkflowCompleted(ByVal sender As Object, _
                  ByVal e As WorkflowCompletedEventArgs)
            If Me.result.InvokeRequired Then
                Me.result.Invoke(New EventHandler(Of WorkflowCompletedEventArgs) _
                    (AddressOf Me.workflowRuntime_WorkflowCompleted), _
                    New Object() {sender, e})
            Else
                Me.result.Text = e.OutputParameters("Result").ToString()

                ' Clear fields
                Me.amount.Text = String.Empty

                ' Disable buttons
                Me.approveButton.Enabled = False
                Me.rejectButton.Enabled = False
            End If
        End Sub

        Private Sub approveButton_Click(ByVal sender As Object, ByVal e As EventArgs) _
                  Handles approveButton.Click
        End Sub

        Private Sub rejectButton_Click(ByVal sender As Object, ByVal e As EventArgs) _
                  Handles rejectButton.Click
        End Sub

        Private Sub amount_KeyPress(ByVal sender As Object, _
                  ByVal e As KeyPressEventArgs) Handles amount.KeyPress
            If (Not Char.IsControl(e.KeyChar)) AndAlso _
                ((Not Char.IsDigit(e.KeyChar))) Then
                e.KeyChar = Char.MinValue
            End If
        End Sub

        Private Sub amount_TextChanged(ByVal sender As Object, ByVal e As EventArgs) _
                  Handles amount.TextChanged
            submitButton.Enabled = amount.Text.Length > 0
        End Sub
    End Class

    Friend Class Program
        Private Sub New()
        End Sub
        <STAThread()> _
        Shared Sub Main()
            Application.EnableVisualStyles()
            Application.Run(New MainForm())
        End Sub
    End Class
End Namespace

File: WWF/Tutorials/Tutorial1/Exercise2/Task1/CS/SimpleExpenseReport/Program.cs
//---------------------------------------------------------------------
//  This file is part of the Windows Workflow Foundation SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------
using System;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;
using System.Collections.Generic;
using System.Workflow.Runtime;
using System.Workflow.Runtime.Hosting;
using System.Workflow.Activities;
using System.Threading;

namespace Microsoft.Samples.Workflow.Tutorials.SequentialWorkflow
{
    public class MainForm : Form
    {
        private System.Windows.Forms.Label label1;
        private System.Windows.Forms.TextBox result;
        private System.Windows.Forms.Label label2;
        private System.Windows.Forms.Button submitButton;
        private System.Windows.Forms.Label approvalState;
        private System.Windows.Forms.Button approveButton;
        private System.Windows.Forms.Button rejectButton;
        private System.Windows.Forms.TextBox amount;
        private System.Windows.Forms.Panel panel1;

        private System.ComponentModel.IContainer components = null;

        private WorkflowRuntime workflowRuntime = null;
        private WorkflowInstance workflowInstance = null;

        public MainForm()
        {
            InitializeComponent();

            // Collapse approve/reject panel
            this.Height = this.MinimumSize.Height;

            this.workflowRuntime = new WorkflowRuntime();
            workflowRuntime.StartRuntime();

            workflowRuntime.WorkflowCompleted += new 
                EventHandler<WorkflowCompletedEventArgs>
                (workflowRuntime_WorkflowCompleted);
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing && (components != null))
            {
                components.Dispose();
            }
            base.Dispose(disposing);
        }

        private void InitializeComponent()
        {
            this.label1 = new System.Windows.Forms.Label();
            this.result = new System.Windows.Forms.TextBox();
            this.label2 = new System.Windows.Forms.Label();
            this.submitButton = new System.Windows.Forms.Button();
            this.approvalState = new System.Windows.Forms.Label();
            this.approveButton = new System.Windows.Forms.Button();
            this.rejectButton = new System.Windows.Forms.Button();
            this.amount = new System.Windows.Forms.TextBox();
            this.panel1 = new System.Windows.Forms.Panel();
            this.panel1.SuspendLayout();
            this.SuspendLayout();
            // 
            // label1
            // 
            this.label1.AutoSize = true;
            this.label1.Location = new System.Drawing.Point(10, 13);
            this.label1.Name = "label1";
            this.label1.Size = new System.Drawing.Size(43, 13);
            this.label1.TabIndex = 1;
            this.label1.Text = "Amount";
            // 
            // result
            // 
            this.result.Location = new System.Drawing.Point(13, 70);
            this.result.Name = "result";
            this.result.ReadOnly = true;
            this.result.Size = new System.Drawing.Size(162, 20);
            this.result.TabIndex = 1;
            this.result.TabStop = false;
            // 
            // label2
            // 
            this.label2.AutoSize = true;
            this.label2.Location = new System.Drawing.Point(10, 54);
            this.label2.Name = "label2";
            this.label2.Size = new System.Drawing.Size(37, 13);
            this.label2.TabIndex = 3;
            this.label2.Text = "Result";
            // 
            // submitButton
            // 
            this.submitButton.Enabled = false;
            this.submitButton.Location = new System.Drawing.Point(56, 95);
            this.submitButton.Name = "submitButton";
            this.submitButton.Size = new System.Drawing.Size(75, 23);
            this.submitButton.TabIndex = 2;
            this.submitButton.Text = "Submit";
            this.submitButton.Click += new System.EventHandler(this.submitButton_Click);
            // 
            // approvalState
            // 
            this.approvalState.AutoSize = true;
            this.approvalState.Location = new System.Drawing.Point(10, 9);
            this.approvalState.Name = "approvalState";
            this.approvalState.Size = new System.Drawing.Size(49, 13);
            this.approvalState.TabIndex = 4;
            this.approvalState.Text = "Approval";
            // 
            // approveButton
            // 
            this.approveButton.Enabled = false;
            this.approveButton.Location = new System.Drawing.Point(11, 25);
            this.approveButton.Name = "approveButton";
            this.approveButton.Size = new System.Drawing.Size(75, 23);
            this.approveButton.TabIndex = 0;
            this.approveButton.Text = "Approve";
            this.approveButton.Click += 
                new System.EventHandler(this.approveButton_Click);
            // 
            // rejectButton
            // 
            this.rejectButton.Enabled = false;
            this.rejectButton.Location = new System.Drawing.Point(86, 25);
            this.rejectButton.Name = "rejectButton";
            this.rejectButton.Size = new System.Drawing.Size(75, 23);
            this.rejectButton.TabIndex = 1;
            this.rejectButton.Text = "Reject";
            this.rejectButton.Click += 
                new System.EventHandler(this.rejectButton_Click);
            // 
            // amount
            // 
            this.amount.Location = new System.Drawing.Point(13, 29);
            this.amount.MaxLength = 9;
            this.amount.Name = "amount";
            this.amount.Size = new System.Drawing.Size(162, 20);
            this.amount.TabIndex = 0;
            this.amount.KeyPress += 
                new System.Windows.Forms.KeyPressEventHandler(this.amount_KeyPress);
            this.amount.TextChanged += 
                new System.EventHandler(this.amount_TextChanged);
            // 
            // panel1
            // 
            this.panel1.Controls.Add(this.approvalState);
            this.panel1.Controls.Add(this.approveButton);
            this.panel1.Controls.Add(this.rejectButton);
            this.panel1.Location = new System.Drawing.Point(7, 124);
            this.panel1.Name = "panel1";
            this.panel1.Size = new System.Drawing.Size(172, 66);
            this.panel1.TabIndex = 8;
            // 
            // MainForm
            // 
            this.AcceptButton = this.submitButton;
            this.AutoScaleDimensions = new System.Drawing.SizeF(6F, 13F);
            this.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Font;
            this.ClientSize = new System.Drawing.Size(187, 201);
            this.MinimumSize = new System.Drawing.Size(197, 155);
            this.Controls.Add(this.result);
            this.Controls.Add(this.label2);
            this.Controls.Add(this.panel1);
            this.Controls.Add(this.amount);
            this.Controls.Add(this.submitButton);
            this.Controls.Add(this.label1);
            this.FormBorderStyle = System.Windows.Forms.FormBorderStyle.Fixed3D;
            this.MaximizeBox = false;
            this.MinimizeBox = false;
            this.Name = "MainForm";
            this.Text = "Simple Expense Report";
            this.panel1.ResumeLayout(false);
            this.panel1.PerformLayout();
            this.ResumeLayout(false);
            this.PerformLayout();
        }

        private void submitButton_Click(object sender, EventArgs e)
        {
            Type type = typeof(ExpenseReportWorkflow);

            // Construct workflow parameters
            Dictionary<string, object> properties = new Dictionary<string, object>();
            properties.Add("Amount", Int32.Parse(this.amount.Text));

            // Start the workflow
            workflowInstance = workflowRuntime.CreateWorkflow(type, properties);
            workflowInstance.Start();
        }

        void workflowRuntime_WorkflowCompleted(object sender, 
            WorkflowCompletedEventArgs e)
        {
            if (this.result.InvokeRequired)
            {
                this.result.Invoke(new EventHandler<WorkflowCompletedEventArgs>
                    (this.workflowRuntime_WorkflowCompleted), sender, e);
            }
            else
            {
                this.result.Text = e.OutputParameters["Result"].ToString();

                // Clear fields
                this.amount.Text = string.Empty;

                // Disable buttons
                this.approveButton.Enabled = false;
                this.rejectButton.Enabled = false;
            }
        }

        private void approveButton_Click(object sender, EventArgs e)
        {
        }

        private void rejectButton_Click(object sender, EventArgs e)
        {
        }

        private void amount_KeyPress(object sender, KeyPressEventArgs e)
        {
            if (!Char.IsControl(e.KeyChar) && (!Char.IsDigit(e.KeyChar)))
                e.KeyChar = Char.MinValue;            
        }

        private void amount_TextChanged(object sender, EventArgs e)
        {
            submitButton.Enabled = amount.Text.Length > 0;
        }
    }

static class Program
    {
        [STAThread]
        static void Main()
        {
            Application.EnableVisualStyles();
            Application.Run(new MainForm());
        }
    }
}
